src/ILOManager.sol

```diff
@@ -2,253 +2,396 @@
         _project.platformFee = platformFee;
         _project.performanceFee = performanceFee;
         emit FeesForProjectSet(projectId, platformFee, performanceFee);
     }
 
+    function setILOSalePoolImplementation(
+        address iloSalePoolImplementation
+    ) external override onlyOwner {
+        emit SalePoolImplementationChanged(
+            ILO_POOL_SALE_IMPLEMENTATION,
+            iloSalePoolImplementation
+        );
+        ILO_POOL_SALE_IMPLEMENTATION = iloSalePoolImplementation;
+    }
+
+    function project(
+        string calldata projectId
+    ) external view override returns (Project memory) {
+        return _projects[projectId];
+    }
+
     /// @inheritdoc IILOManager
-    function feesForProject(string calldata projectId) external override view returns (uint16, uint16) {
+    function feesForProject(
+        string calldata projectId
+    ) external view override returns (uint16, uint16) {
         Project storage _project = _projects[projectId];
         return (_project.platformFee, _project.performanceFee);
     }
+
+    function _deployIloPool(
+        IILOPoolBase.InitPoolParams calldata params,
+        address implementation
+    ) internal returns (address deployedAddress) {
+        // dont need to check if project is exist because only project admin can call this function
+        Project storage _project = _projects[params.projectId];
+        _checkTicks(params.tickLower, params.tickUpper, _project.fee);
+        uint256 projectNonce = ++_project.nonce;
+
+        // this salt make sure that pool address can not be represented in any other chains
+        bytes32 salt = keccak256(
+            abi.encodePacked(ChainId.get(), params.projectId, projectNonce)
+        );
+        deployedAddress = Clones.cloneDeterministic(implementation, salt);
+        emit ILOPoolCreated(params.projectId, deployedAddress);
+    }
+
+    function _initUniV3PoolIfNecessary(
+        PoolAddress.PoolKey memory poolKey,
+        uint160 sqrtPriceX96
+    ) internal returns (address pool) {
+        pool = IUniswapV3Factory(UNIV3_FACTORY).getPool(
+            poolKey.token0,
+            poolKey.token1,
+            poolKey.fee
+        );
+        if (pool == address(0)) {
+            pool = IUniswapV3Factory(UNIV3_FACTORY).createPool(
+                poolKey.token0,
+                poolKey.token1,
+                poolKey.fee
+            );
+            IUniswapV3Pool(pool).initialize(sqrtPriceX96);
+        } else {
+            (uint160 sqrtPriceX96Existing, , , , , , ) = IUniswapV3Pool(pool)
+                .slot0();
+            if (sqrtPriceX96Existing == 0) {
+                IUniswapV3Pool(pool).initialize(sqrtPriceX96);
+            } else {
+                require(sqrtPriceX96Existing == sqrtPriceX96, 'UV3P');
+            }
+        }
+    }
+
+    function _cancelProject(Project storage _project) internal {
+        _project.status = ProjectStatus.CANCELLED;
```

Comment:
```suggestion
        require(_project.status == ProjectStatus.INITIALIZED, 'NA');
        _project.status = ProjectStatus.CANCELLED;
```

Reply:
dont need to check if project is canceled or not because when project is canceled, all of its pools are canceled and calling pool cancel() below will revert